File: 817. Linked List Components.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def numComponents(self, head: ListNode, G: List[int]) -> int:
        setG = set(G)
        res = 0
        node = head
        while node:
            if node.val in setG and (not node.next or node.next.val not in setG):
                res += 1
            node = node.next
        return res

File: 817. Linked List Components.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public int numComponents(ListNode head, int[] G) {
        Set<Integer> s = new HashSet<Integer>();
        for (int i : G) {
            s.add(i);
        }
        int ans = 0;
        while (head != null) {
            if (s.contains(head.val) && (head.next == null || !s.contains(head.next.val))) {ans++;}
            head = head.next;
        }
        return ans;
    }
}